File: Problem2_SearchRotatedSortedArray.cpp
// Time Complexity : O(log(n))
// Space Complexity : O(n)
// Did this code successfully run on Leetcode :yes  
// Any problem you faced while coding this : no


// Your code here along with comments explaining your approach

class Solution {
public:
    int search(vector<int>& nums, int target) {
        
        if(nums.empty()||nums.size()==0) return -1;
        int n = nums.size();
        int low =0;
        int high=n-1;
        
        while(low<=high){
            int mid=low+(high-low)/2;
            if(nums[mid]==target) return mid;
            else if(nums[low]<=nums[mid])
            {   if(nums[low]<=target && target<nums[mid])
                    high=mid-1;
                else
                    low=mid+1;
            }
            
            
            
                
            else{ // right side is sorted
                if(nums[mid]<target && target<=nums[high]){
                    low=mid+1;
                }
                else
                    high=mid-1;                            
            }
        }
            
        return -1;
        
    }
};

File: Problem2_SearchRotatedSortedArray.py
# Time Complexity: O(log n)
# Space Complexity: O(1)
# Did this code successfully run on Leetcode: Yes

# Solution:
"""
1. Using binary search approach, find the mid element and check if any of the sides is sorted.
2. If target lies within the range of the sorted side, perform binary search in that range.
3. Else perform the search in the unsorted side. Repeat the steps until the element is found.
"""

class Solution:
    def search(self, nums: List[int], target: int) -> int:
        if not nums or len(nums) == 0:
            return -1

        low = 0
        high = len(nums) - 1

        while low <= high:
            mid = low + (high - low) // 2

            if target == nums[mid]:
                return mid

            # Left side is sorted
            if nums[low] <= nums[mid]:
                if nums[low] <= target < nums[mid]:
                    high = mid - 1
                else:
                    low = mid + 1

            # Right side is sorted
            else:
                if nums[mid] < target <= nums[high]:
                    low = mid + 1
                else:
                    high = mid - 1

        return -1